wiki/C++-Style-Guide.md

```diff
@@ -0,0 +1,814 @@
+```c++
+class MyClass : public MyBaseClass { // Brace on the same line.
+public: // Do not indent the tags for "public," etc.
+    MyClass(); // Default constructor always comes first.
+    explicit MyClass(int size); // Avoid unexpected implicit conversion.
+    
+    // These comments summarize the following block of functions, with enough
+    // detail for anyone using this interface (but who does not care about the
+    // implementation details).
+    bool IsEmpty() const; // Use "const" whenever possible.
+    size_t Size() const; // Use unsigned types for sizes.
+    
+    // Another set of descriptive comments.
+    void SetLabel(int index, const std::string &label);
+    const std::string &GetLabel(int index) const;
+    void GetIndex(const std::string &label) const;
+    
+    // This class behaves like an array; you can look up elements by index, or
+    // by label (if a given label has been assigned to an index).
+    double operator[](int index) const; // Only use obvious operator overloads.
+    double operator[](const std::string &label) const;
+    
+    
+private: // Two blank lines separating each section.
+    // Clear any index previously associated with the given label.
+    void ClearLabel(const std::string &label);
+    
+    
+private: // Separate sections for functions and for variables.
+    std::vector<double> values;
+    std::vector<std::string> labels;
+    std::map<std::string, size_t> indices;
+};
+```
+
+## Inline functions
+
+Do not define any functions within the class definition.
+The class definition should provide a concise definition of the interface of the class, so that someone using the class can scan through it quickly. Defining functions inline makes this more difficult.
+If you have determined that making a function inline will substantially improve performance, or if it is a templated function, define it in the header after the end of the class definition.
+
+## Initialization order
+
+Initializer lists in constructors should list the variables in the same order they occur in the class.
+Variables are initialized in the order they occur in the class definition, not in the initializer list. Most compilers will report a warning if you change the order because it means your initializer list might be written to depend on a variable that has not been initialized yet.
+
+## Explicit constructors
+
+Constructors with only one argument should be labeled "explicit".
+This is to avoid defining implicit type conversions that might then be used by the compiler in surprising ways.
+
+## Operator overloading
+
+Use operator overloading only if implementing an arithmetic class and only if the operator has its well-established mathematical sense.
+For example, the "Point" class can define Point + Point, Point * scalar, etc., but it should not define Point * Point because it is ambiguous whether that should be an element-by-element product, a dot product, or a cross product.
```

Comment:
```suggestion
For example, the `Point` class can define Point + Point, Point * scalar, etc., but it should not define Point * Point because it is ambiguous whether that should be an element-by-element product, a dot product, or a cross product.
```